File: solutions/haskell/acronym/2/src/Acronym.hs
module Acronym (abbreviate) where

import Data.Char (toUpper, isUpper)

abbreviate :: String -> String
abbreviate str = concat $ map initials $ tokenize str
  where
    initials :: String -> String
    initials s
      | allUpper == []                  = [toUpper (head s)]
      | (length allUpper) == (length s) = [head s]
      | otherwise                       = allUpper
      where
        allUpper = filter isUpper s
    tokenize :: String -> [String]
    tokenize s = words $ map replace s
    replace :: Char -> Char
    replace x = case x of
                  '-' -> ' '
                  c -> c

File: solutions/erlang/acronym/2/src/acronym.erl
-module(acronym).
-import(string, [tokens/2, to_upper/1]).
-import(lists, [map/2]).
-export([abbreviate/1]).

-spec abbreviate(string()) -> string().
abbreviate(Phrase) ->
  to_upper(
    map(
      fun hd/1,
      tokens(Phrase, " -_"))).
